### DISCUSSION TITLE
Can I use the NPM parser to parse TypeScript?

### DISCUSSION THREAD

**Discussion by Sleepful** [oxc-project/oxc#3108] [Q&A] [answered]
Hello, I found: https://www.npmjs.com/package/oxc-parser

It works well to parse JS files, but it is giving me errors with TS files. Is this expected? It seems the rust parser (https://github.com/oxc-project/oxc/blob/main/crates/oxc_parser/examples/parser.rs) works fine with TS files, so I was expecting the NPM parser to work well too.

**Comment by Boshen** [comment]
Try passing in the filename with the correct extension, i.e. `test.ts`.

**Reply by Sleepful** [reply]
How so? If I understood correctly, `parseAsync` param is going to be the string source code. It does not take in a filename, does it?

E.g.
```ts
oxc.parseAsync(`import { Button } from "../components/Button.tsx";`)
// works fine
oxc.parseAsync(`import type { Signal } from "@preact/signals";`)
// this one gives error because it is not valid JavaScript
```

**Reply by Sleepful** [reply]
The above always return the "javascript" language:

```
{
  type: "Program",
  start: 0,
  end: 7,
  sourceType: {
    language: "javascript",
    moduleKind: "script",
    variant: "standard",
    alwaysStrict: false
  }
```

Passing in a filename does not work.

```
oxc.parseAsync(`test.ts`) // does not actually read the file
```

**Reply by Boshen** [reply]
Does the TypeScript definition not work?

```
export function parseSync(sourceText: string, options?: ParserOptions | undefined | null): ParseResult
```

I'll update the README example regarding the options.

**Reply by Sleepful** [reply, marked as answer]
@Boshen oh, you are right. I see what you mean now.

This works fine:

```ts
parseSync(`import type { Signal } from "@preact/signals";`, {
  sourceFilename: "some.ts",
})
```
Now that I remember, I _had_ looked at https://github.com/oxc-project/oxc/blob/main/napi/parser/index.d.ts#L82 before posting this discussion, but I was looking for something like "filetype" that I could tell it "TypeScript". It didn't occur to me that `sourceFilename` was important! Adding a sample in the README for the NPM package where `sourceFilename` is used for a TS file would have been enough.
